tests/test_scheduler.py

```diff
@@ -0,0 +1,345 @@
+            priorities.append(self.scheduler.next_request().priority)
+
+        self.assertEqual(priorities,
+                         sorted([x[1] for x in _PRIORITIES], key=lambda x: -x))
+
+
+class TestSchedulerInMemory(BaseSchedulerInMemoryTester, unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.ScrapyPriorityQueue'
+
+
+class TestSchedulerOnDisk(BaseSchedulerOnDiskTester, unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.ScrapyPriorityQueue'
+
+
+_SLOTS = [("http://foo.com/a", 'a'),
+          ("http://foo.com/b", 'a'),
+          ("http://foo.com/c", 'b'),
+          ("http://foo.com/d", 'b'),
+          ("http://foo.com/e", 'c'),
+          ("http://foo.com/f", 'c')]
+
+
+class TestMigration(unittest.TestCase):
+
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp()
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+    def _migration(self, tmp_dir):
+        prev_scheduler_handler = SchedulerHandler()
+        prev_scheduler_handler.priority_queue_cls = 'scrapy.pqueues.ScrapyPriorityQueue'
+        prev_scheduler_handler.jobdir = tmp_dir
+
+        prev_scheduler_handler.create_scheduler()
+        for url in _URLS:
+            prev_scheduler_handler.scheduler.enqueue_request(Request(url))
+        prev_scheduler_handler.close_scheduler()
+
+        next_scheduler_handler = SchedulerHandler()
+        next_scheduler_handler.priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+        next_scheduler_handler.jobdir = tmp_dir
+
+        next_scheduler_handler.create_scheduler()
+
+    def test_migration(self):
+        with self.assertRaises(ValueError):
+            self._migration(self.tmpdir)
+
+
+class TestSchedulerWithDownloaderAwareInMemory(BaseSchedulerInMemoryTester,
+                                               unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+
+    def test_logic(self):
+        downloader = self.mock_crawler.engine.downloader
+        for url, slot in _SLOTS:
+            request = Request(url)
+            downloader._set_slot_key(slot, request, None)
+            self.scheduler.enqueue_request(request)
+
+        slots = list()
+        requests = list()
+        while self.scheduler.has_pending_requests():
+            request = self.scheduler.next_request()
+            slot = downloader._get_slot_key(request, None)
+            slots.append(slot)
+            downloader.increment(slot)
+            requests.append(request)
+        self.assertEqual(len(slots), len(_SLOTS))
+
+        for request in requests:
+            slot = downloader._get_slot_key(request, None)
+            self.mock_crawler.engine.downloader.decrement(slot)
+
+        unique_slots = len(set(s for _, s in _SLOTS))
+        for i in range(0, len(_SLOTS), unique_slots):
+            part = slots[i:i + unique_slots]
+            self.assertEqual(len(part), len(set(part)))
+
+
+def _is_slots_unique(base_slots, result_slots):
```

Comment:
Hm, I see what this function is checking; this check makes sense. But it was not easy to understand, and the check has caveats - e.g. it will fail for any algorithm (including good algorithms) for some input slots, e.g. if you have 1 request for slot A and 5 requests for slot B; it only works if there is roughly the same number of requests (+-1?) for each slot in the input data.

So I think it'd be good to 

a) rename it - e.g. `_assert_fair`, `_assert_round_robin` (though this one is not precise)?, and
b) add either a comment or an assert about kind of input data this check works on properly.